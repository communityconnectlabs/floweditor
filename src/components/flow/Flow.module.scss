@import "variables.module.scss";

.empty_flow {
  margin: 13px;
  width: 400px;

  h1 {
    font-family: "Helvetica Neue", "RobotoThin", sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}

.contextMenu {
  z-index: 150;
  border: 1px solid  #ccc;
  background-color: hsl(0,0%,100%);
  border-radius: 5px;
  .menuItem {
    font-size: 14px;
    border-radius: 4px;
    background-color: hsl(0,0%,100%);
    outline: none;
    padding: 10px 16px;
    &:hover {
      color: white;
      background-color: #289f9b;
    }
  }
}
