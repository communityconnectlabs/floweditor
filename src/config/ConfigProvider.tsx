/* eslint-disable @typescript-eslint/no-explicit-any */
import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType
  };

  constructor(props: ConfigProviderProps) {
    super(props);
    const win = window as any;
    if (win.isMobile && win.isMobile()) {
      this.props.config.mutable = false;
    }

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
