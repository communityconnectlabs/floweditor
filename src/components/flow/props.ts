import { UpdateLocalizations } from 'components/nodeeditor/NodeEditor';
import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';
import { CompletionSchema } from 'utils/completion';

export interface ActionFormProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;
  completionSchema: CompletionSchema;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
  onClose(canceled: boolean): void;
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};
