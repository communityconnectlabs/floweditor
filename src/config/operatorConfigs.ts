import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES, ONLINE } from 'config/interfaces';

export const intentOperatorList: Operator[] = [
  {
    type: Operators.has_top_intent,
    verboseName: 'has top intent',
    operands: 2,
    visibility: ONLINE
  },
  {
    type: Operators.has_intent,
    verboseName: 'has intent',
    operands: 2,
    visibility: ONLINE
  }
];

export const operatorConfigList: Operator[] = [
  {
    type: Operators.has_any_word,
    verboseName: 'has any of the words',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_all_words,
    verboseName: 'has all of the words',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_phrase,
    verboseName: 'has the phrase',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_only_phrase,
    verboseName: 'has only the phrase',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_beginning,
    verboseName: 'starts with',
    operands: 1
  },
  {
    type: Operators.has_text,
    verboseName: 'has some text',
    operands: 0,
    categoryName: 'Has Text',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_number,
    verboseName: 'has a number',
    operands: 0,
    categoryName: 'Has Number'
  },
  {
    type: Operators.has_number_between,
    verboseName: 'has a number between',
    operands: 2
  },
  {
    type: Operators.has_number_lt,
    verboseName: 'has a number below',
    operands: 1
  },
  {
    type: Operators.has_number_lte,
    verboseName: 'has a number at or below',
    operands: 1
  },
  {
    type: Operators.has_number_eq,
    verboseName: 'has a number equal to',
    operands: 1
  },
  {
    type: Operators.has_number_gte,
    verboseName: 'has a number at or above',
    operands: 1
  },
  {
    type: Operators.has_number_gt,
    verboseName: 'has a number above',
    operands: 1
  },
  {
    type: Operators.has_date,
    verboseName: 'has a date',
    operands: 0,
    categoryName: 'Has Date',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_lt,
    verboseName: 'has a date before',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_eq,
    verboseName: 'has a date equal to',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_gt,
    verboseName: 'has a date after',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_time,
    verboseName: 'has a time',
    operands: 0,
    categoryName: 'Has Time',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_group,
    verboseName: 'is in the group',
    operands: 1,
    visibility: HIDDEN
  },
  {
    type: Operators.has_category,
    verboseName: 'has the category',
    operands: 0,
    visibility: HIDDEN
  },
  {
    type: Operators.has_phone,
    verboseName: 'has a phone number',
    operands: 0,
    categoryName: 'Has Phone'
  },
  {
    type: Operators.has_email,
    verboseName: 'has an email',
    operands: 0,
    categoryName: 'Has Email',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_state,
    verboseName: 'has state',
    operands: 0,
    categoryName: 'Has State'
  },
  {
    type: Operators.has_district,
    verboseName: 'has district',
    operands: 1,
    categoryName: 'Has District'
  },
  {
    type: Operators.has_ward,
    verboseName: 'has ward',
    operands: 2,
    categoryName: 'Has Ward'
  },
  {
    type: Operators.has_error,
    verboseName: 'has an error',
    operands: 0,
    categoryName: 'Has Error',
    visibility: HIDDEN
  },
  {
    type: Operators.has_value,
    verboseName: 'is not empty',
    operands: 0,
    categoryName: 'Not Empty'
  },
  {
    type: Operators.has_pattern,
    verboseName: 'matches regex',
    operands: 1
  }
];

export const operatorConfigMap: OperatorMap = [...operatorConfigList, ...intentOperatorList].reduce(
  (map: OperatorMap, operatorConfig: Operator) => {
    map[operatorConfig.type] = operatorConfig;
    return map;
  },
  {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
