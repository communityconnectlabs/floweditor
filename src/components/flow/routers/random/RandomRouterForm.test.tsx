import { RouterFormProps } from 'components/flow/props';
import { getTypeConfig } from 'config';
import { Types } from 'config/interfaces';
import * as React from 'react';
import { fireEvent, render } from 'test/utils';
import { composeComponentTestUtils, mock } from 'testUtils';
import {
  createRandomNode,
  createRenderNode,
  createSendMsgAction,
  getRouterFormProps
} from 'testUtils/assetCreators';
import * as utils from 'utils';

import RandomRouterForm from './RandomRouterForm';

mock(utils, 'createUUID', utils.seededUUIDs());

const { setup } = composeComponentTestUtils<RouterFormProps>(
  RandomRouterForm,
  getRouterFormProps(createRandomNode(2))
);

describe(RandomRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should initialize existing random', () => {
    const props: RouterFormProps = {
      nodeSettings: {
        originalNode: createRandomNode(5)
      },
      typeConfig: getTypeConfig(Types.split_by_random),
      assetStore: null,
      updateRouter: jest.fn(),
      onTypeChange: jest.fn(),
      onClose: jest.fn()
    };

    const { baseElement } = render(<RandomRouterForm {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should remove exits when shrinking', () => {
    const props: RouterFormProps = {
      nodeSettings: {
        originalNode: createRandomNode(5)
      },
      typeConfig: getTypeConfig(Types.split_by_random),
      assetStore: null,
      updateRouter: jest.fn(),
      onTypeChange: jest.fn(),
      onClose: jest.fn()
    };

    const { baseElement, getAllByTestId, getByDisplayValue, getByText } = render(
      <RandomRouterForm {...props} />
    );

    // we start off with five input boxes for our buckets
    expect(getAllByTestId('input').length).toEqual(5);

    // the second select box is our bucket, choose 3 buckets
    fireEvent.change(getAllByTestId('select')[1], {
      target: { value: 3 }
    });

    // now we should only have three input buckets
    expect(getAllByTestId('input').length).toEqual(3);
    expect(baseElement).toMatchSnapshot();

    // now lets save our form
    fireEvent.click(getByText('Ok'));

    expect(props.updateRouter).toMatchCallSnapshot();
  });

  it('should convert from a non-random node', () => {
    const { wrapper, instance, props } = setup(true, {
      nodeSettings: {
        $set: {
          originalAction: createSendMsgAction({ text: '' }),
          originalNode: createRenderNode({
            actions: [createSendMsgAction({ text: 'A message' })],
            exits: [{ uuid: utils.createUUID() }],
            ui: {
              type: Types.execute_actions,
              position: { left: 100, top: 100 }
            }
          })
        }
      }
    });

    instance.handleSave();

    // our orginal node should still only have one exit
    expect(props.nodeSettings.originalNode.node.exits.length).toBe(1);

    expect(wrapper).toMatchSnapshot();
    expect(props.updateRouter).toMatchCallSnapshot();
    expect(props.nodeSettings.originalNode).toMatchSnapshot();
  });
});
