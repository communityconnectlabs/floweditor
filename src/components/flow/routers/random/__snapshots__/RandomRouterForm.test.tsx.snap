// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomRouterForm should convert from a non-random node 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="lead_in"
  >
    Split them randomly into one of
  </div>
  <div
    className="bucket_select"
  >
    <SelectElement
      entry={
        Object {
          "value": Object {
            "label": "2 buckets",
            "value": "2",
          },
        }
      }
      name="Buckets"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "2 buckets",
            "value": "2",
          },
          Object {
            "label": "3 buckets",
            "value": "3",
          },
          Object {
            "label": "4 buckets",
            "value": "4",
          },
          Object {
            "label": "5 buckets",
            "value": "5",
          },
          Object {
            "label": "6 buckets",
            "value": "6",
          },
          Object {
            "label": "7 buckets",
            "value": "7",
          },
          Object {
            "label": "8 buckets",
            "value": "8",
          },
          Object {
            "label": "9 buckets",
            "value": "9",
          },
          Object {
            "label": "10 buckets",
            "value": "10",
          },
        ]
      }
      styles={
        Object {
          "control": [Function],
          "indicatorsContainer": [Function],
          "input": [Function],
          "menuList": [Function],
          "option": [Function],
          "singleValue": [Function],
          "valueContainer": [Function],
        }
      }
    />
  </div>
  <div
    className="bucket_list"
  >
    <Connect(TextInputElement)
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="306dee46-46c7-4a8a-8a33-ab2b7e42b02e"
      name="306dee46-46c7-4a8a-8a33-ab2b7e42b02e"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="90e20a49-6ecc-4b0b-97bf-bcbd37069b63"
      name="90e20a49-6ecc-4b0b-97bf-bcbd37069b63"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`RandomRouterForm should convert from a non-random node 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "b5292cef-6113-4066-80f5-371178abf28b",
        },
        Object {
          "destination_uuid": null,
          "uuid": "f627a69c-a227-451f-9520-ce969f02162e",
        },
      ],
      "router": Object {
        "categories": Array [
          Object {
            "exit_uuid": "b5292cef-6113-4066-80f5-371178abf28b",
            "name": "Bucket 1",
            "uuid": "306dee46-46c7-4a8a-8a33-ab2b7e42b02e",
          },
          Object {
            "exit_uuid": "f627a69c-a227-451f-9520-ce969f02162e",
            "name": "Bucket 2",
            "uuid": "90e20a49-6ecc-4b0b-97bf-bcbd37069b63",
          },
        ],
        "type": "random",
      },
      "uuid": "c99dcf07-2ad6-4730-9e99-810a39f5a58d",
    },
    "ui": Object {
      "config": null,
      "position": null,
      "type": "split_by_random",
    },
  },
]
`;

exports[`RandomRouterForm should convert from a non-random node 3`] = `
Object {
  "inboundConnections": Object {},
  "node": Object {
    "actions": Array [
      Object {
        "all_urns": false,
        "text": "A message",
        "type": "send_msg",
        "uuid": "38282c32-100e-4376-af57-7f6a36ddafba",
      },
    ],
    "exits": Array [
      Object {
        "uuid": "b5a6cc81-02ba-4931-b4be-a0d7204be029",
      },
    ],
    "uuid": "c99dcf07-2ad6-4730-9e99-810a39f5a58d",
  },
  "ui": Object {
    "position": Object {
      "left": 100,
      "top": 100,
    },
    "type": "execute_actions",
  },
}
`;

exports[`RandomRouterForm should initialize existing random 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header split_by_random"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split Randomly
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <select
              data-testid="select"
            >
              <option
                value="send_msg"
              >
                Send the contact a message
              </option>
              <option
                value="wait_for_response"
              >
                Wait for the contact to respond
              </option>
              <option
                value="send_broadcast"
              >
                Send somebody else a message
              </option>
              <option
                value="add_input_labels"
              >
                Label the incoming message
              </option>
              <option
                value="add_contact_urn"
              >
                Add a URN for the contact
              </option>
              <option
                value="add_contact_groups"
              >
                Add the contact to a group
              </option>
              <option
                value="remove_contact_groups"
              >
                Remove the contact from a group
              </option>
              <option
                value="set_contact_field"
              >
                Update the contact
              </option>
              <option
                value="send_email"
              >
                Send an email
              </option>
              <option
                value="set_run_result"
              >
                Save a result for this flow
              </option>
              <option
                value="call_webhook"
              >
                Call a webhook
              </option>
              <option
                value="call_resthook"
              >
                Call Zapier
              </option>
              <option
                value="enter_flow"
              >
                Enter another flow
              </option>
              <option
                value="start_session"
              >
                Start somebody else in a flow
              </option>
              <option
                value="transfer_airtime"
              >
                Send the contact airtime
              </option>
              <option
                value="split_by_intent"
              >
                Split by intent
              </option>
              <option
                value="split_by_expression"
              >
                Split by a custom expression
              </option>
              <option
                value="split_by_contact_field"
              >
                Split by a contact field
              </option>
              <option
                value="split_by_run_result"
              >
                Split by a result in the flow
              </option>
              <option
                value="split_by_random"
              >
                Split by random chance
              </option>
              <option
                value="split_by_groups"
              >
                Split by group membership
              </option>
            </select>
          </div>
        </div>
        <div
          class="lead_in"
        >
          Split them randomly into one of
        </div>
        <div
          class="bucket_select"
        >
          <div
            class="ele"
          >
            <select
              data-testid="select"
            >
              <option
                value="2"
              >
                2 buckets
              </option>
              <option
                value="3"
              >
                3 buckets
              </option>
              <option
                value="4"
              >
                4 buckets
              </option>
              <option
                value="5"
              >
                5 buckets
              </option>
              <option
                value="6"
              >
                6 buckets
              </option>
              <option
                value="7"
              >
                7 buckets
              </option>
              <option
                value="8"
              >
                8 buckets
              </option>
              <option
                value="9"
              >
                9 buckets
              </option>
              <option
                value="10"
              >
                10 buckets
              </option>
            </select>
          </div>
        </div>
        <div
          class="bucket_list"
        >
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="e67ce051-f3e4-4c42-807b-e393df5fa382"
                type="text"
                value="Bucket 1"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="e0bc3776-792f-49d2-b4be-48be2cf8159a"
                type="text"
                value="Bucket 2"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="6c22884b-0e35-4206-982e-18320691eda9"
                type="text"
                value="Bucket 3"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="a62e8284-49f8-4292-adbc-d5f6607f85c5"
                type="text"
                value="Bucket 4"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="0c8c9239-9b59-473a-a229-d8a9693be270"
                type="text"
                value="Bucket 5"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`RandomRouterForm should remove exits when shrinking 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header split_by_random"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split Randomly
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <select
              data-testid="select"
            >
              <option
                value="send_msg"
              >
                Send the contact a message
              </option>
              <option
                value="wait_for_response"
              >
                Wait for the contact to respond
              </option>
              <option
                value="send_broadcast"
              >
                Send somebody else a message
              </option>
              <option
                value="add_input_labels"
              >
                Label the incoming message
              </option>
              <option
                value="add_contact_urn"
              >
                Add a URN for the contact
              </option>
              <option
                value="add_contact_groups"
              >
                Add the contact to a group
              </option>
              <option
                value="remove_contact_groups"
              >
                Remove the contact from a group
              </option>
              <option
                value="set_contact_field"
              >
                Update the contact
              </option>
              <option
                value="send_email"
              >
                Send an email
              </option>
              <option
                value="set_run_result"
              >
                Save a result for this flow
              </option>
              <option
                value="call_webhook"
              >
                Call a webhook
              </option>
              <option
                value="call_resthook"
              >
                Call Zapier
              </option>
              <option
                value="enter_flow"
              >
                Enter another flow
              </option>
              <option
                value="start_session"
              >
                Start somebody else in a flow
              </option>
              <option
                value="transfer_airtime"
              >
                Send the contact airtime
              </option>
              <option
                value="split_by_intent"
              >
                Split by intent
              </option>
              <option
                value="split_by_expression"
              >
                Split by a custom expression
              </option>
              <option
                value="split_by_contact_field"
              >
                Split by a contact field
              </option>
              <option
                value="split_by_run_result"
              >
                Split by a result in the flow
              </option>
              <option
                value="split_by_random"
              >
                Split by random chance
              </option>
              <option
                value="split_by_groups"
              >
                Split by group membership
              </option>
            </select>
          </div>
        </div>
        <div
          class="lead_in"
        >
          Split them randomly into one of
        </div>
        <div
          class="bucket_select"
        >
          <div
            class="ele"
          >
            <select
              data-testid="select"
            >
              <option
                value="2"
              >
                2 buckets
              </option>
              <option
                value="3"
              >
                3 buckets
              </option>
              <option
                value="4"
              >
                4 buckets
              </option>
              <option
                value="5"
              >
                5 buckets
              </option>
              <option
                value="6"
              >
                6 buckets
              </option>
              <option
                value="7"
              >
                7 buckets
              </option>
              <option
                value="8"
              >
                8 buckets
              </option>
              <option
                value="9"
              >
                9 buckets
              </option>
              <option
                value="10"
              >
                10 buckets
              </option>
            </select>
          </div>
        </div>
        <div
          class="bucket_list"
        >
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="e81b3f1a-0863-4418-8a72-9c9f8aa72b7c"
                type="text"
                value="Bucket 1"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="1888ae1f-2518-4727-98e2-731674e1760e"
                type="text"
                value="Bucket 2"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper"
            >
              <input
                class="textinput"
                data-spec="input"
                data-testid="input"
                maxlength="-1"
                name="dd57500e-a8e8-4a98-8b5b-a33637523a8d"
                type="text"
                value="Bucket 3"
              />
              <div
                class="completion_container hidden"
                data-spec="completion-options"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="options_wrapper"
                >
                  <ul
                    class="option_list"
                    data-spec="completion-list"
                  />
                  <div
                    class="help"
                    data-spec="completion-help"
                  >
                    Tab to complete, enter to select
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`RandomRouterForm should remove exits when shrinking 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
        },
        Object {
          "destination_uuid": null,
          "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
        },
        Object {
          "destination_uuid": null,
          "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        },
      ],
      "router": Object {
        "categories": Array [
          Object {
            "exit_uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
            "name": "Bucket 1",
            "uuid": "e81b3f1a-0863-4418-8a72-9c9f8aa72b7c",
          },
          Object {
            "exit_uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
            "name": "Bucket 2",
            "uuid": "1888ae1f-2518-4727-98e2-731674e1760e",
          },
          Object {
            "exit_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "name": "Bucket 3",
            "uuid": "dd57500e-a8e8-4a98-8b5b-a33637523a8d",
          },
        ],
        "type": "random",
      },
      "uuid": "96249bdc-cb58-4489-a188-0eeb6686a8c9",
    },
    "ui": Object {
      "config": null,
      "position": null,
      "type": "split_by_random",
    },
  },
]
`;

exports[`RandomRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="lead_in"
  >
    Split them randomly into one of
  </div>
  <div
    className="bucket_select"
  >
    <SelectElement
      entry={
        Object {
          "value": Object {
            "label": "2 buckets",
            "value": "2",
          },
        }
      }
      name="Buckets"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "2 buckets",
            "value": "2",
          },
          Object {
            "label": "3 buckets",
            "value": "3",
          },
          Object {
            "label": "4 buckets",
            "value": "4",
          },
          Object {
            "label": "5 buckets",
            "value": "5",
          },
          Object {
            "label": "6 buckets",
            "value": "6",
          },
          Object {
            "label": "7 buckets",
            "value": "7",
          },
          Object {
            "label": "8 buckets",
            "value": "8",
          },
          Object {
            "label": "9 buckets",
            "value": "9",
          },
          Object {
            "label": "10 buckets",
            "value": "10",
          },
        ]
      }
      styles={
        Object {
          "control": [Function],
          "indicatorsContainer": [Function],
          "input": [Function],
          "menuList": [Function],
          "option": [Function],
          "singleValue": [Function],
          "valueContainer": [Function],
        }
      }
    />
  </div>
  <div
    className="bucket_list"
  >
    <Connect(TextInputElement)
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="1dce2b34-9aab-4e20-81c4-3f0408dcb671"
      name="1dce2b34-9aab-4e20-81c4-3f0408dcb671"
      onChange={[Function]}
    />
    <Connect(TextInputElement)
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
      name="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;
